### ISSUE TITLE
What is the big picture plan for 2d charts in griddle?

### ISSUE THREAD

**Issue by samreid** [phetsims/griddle#54]
I saw some issues like https://github.com/phetsims/griddle/issues/50 and https://github.com/phetsims/griddle/issues/53 and thought we should discuss how to move forward with griddle.

The main problem is that there are two implementations, as described in https://github.com/phetsims/griddle/issues/46.  One is capable of scrolling (ScrollingChartNode) and the other is not capable of scrolling (XYPlotNode).

I do not believe we need a "scrolling" implementation and a "non-scrolling" implementation.  A static chart should just be a scrolling chart whose range and domain never change.

The trajectory for https://github.com/phetsims/griddle/issues/46#issuecomment-594173390 was to make ScrollingChartNode more general so that it could support non-scrolling cases easily.  It's usage of ModelViewTransform for managing the data works really well.

I wanted to open this issue now since I saw a lot of work scheduled for XYPlotNode which to me seems like it should be deprecated.

**Comment by samreid** [comment]
From slack, before I wrote up this issue:

@jessegreenberg said:

> Hey Sam! I remembered that you were working on scrolling charts in Griddle. I am working on https://github.com/phetsims/griddle/issues/50 and https://github.com/phetsims/griddle/issues/53 which will change XYPlotNode quite a lot, and probably also add support for scrolling as well. Just thought I would let you know so we don't double up on work. If you would like to discuss further let me know!

@samreid replied:

> I’m glad to hear it.  Let me take a look around and check my status.
> I made significant progress in https://github.com/phetsims/griddle/issues/46 on generalizing ScrollingChartNode
> ScrollingChartNode will support scrolling gridlines for CCK, see https://github.com/phetsims/griddle/issues/31#issuecomment-576784044
> I think we should continue generalizing ScrollingChartNode, rename it to ChartNode or XYPlotNode and eventually eliminate what is now XYPlotNode.
> Or keep moving in the direction of unifying them, so ultimately we have one implementation/api which is capable of scrolling or not scrolling.

@jessegreenberg said:

> Ahh, I see thanks! ScrollingChartNOde is looking good. I like the use of the ModelViewTransform2, I was thinking of changing XYPlotNode to use that instead of its current API.

@samreid said:

> Yes, I agree that is one of its best parts!

@jessegreenberg said:

>> I think we should continue generalizing ScrollingChartNode, rename it to ChartNode or XYPlotNode and eventually eliminate what is now XYPlotNode.

> That sounds great

@samreid said:

> I think we will need more planning and discussion of how & when to get there, but will be good to be taking more steps in that direction.

Assigning @pixelzoom and @ariel-phet for discussion or comment.

**Comment by jessegreenberg** [comment]
> I think we should continue generalizing ScrollingChartNode, rename it to ChartNode or XYPlotNode and eventually eliminate what is now XYPlotNode.

I think this is right. An example of things to do might include 
 - ScrollingChartNode should use GridNode, and the grid should also scroll.
 - Move things out of ScrollingChartNode that should be added with decoration by subtypes (zoom buttons, the "span" node, axis labels)
 - Labels needed for vertical grid lines.
 - ScrollingChartNode was implemented with "time" in mind, perhaps generalize by removing time-specific terminology.
 - XYCursorPlot needs to extend ScrollingCHartNode instead of XYPlotNode.
- ScrollingChartNode is more specifically like a "seismograph", data is plotted from the right side of the graph, which isn't general.
- DynamicSeriesNode has a circular "Pen" where data is plotted, which isn't very general.

**Comment by samreid** [comment]
That all sounds great, thanks!

**Comment by pixelzoom** [comment]
I don't know what the big-picture plan is for griddle.

**Comment by ariel-phet** [comment]
@samreid @jessegreenberg in terms of "big picture" I would say the next big sims on the horizon likely to need support (besides Natural Selection) will be **Moving Man** (likely) and **Fourier: Making Waves** (definitely, since we have funding secure)

Currently, I do not believe there are any design changes planned for the chart aspects of Moving Man, so the Java version should be a good starting point to see if any features are missing. 

@pixelzoom confirmed that Fourier does not currently use JFreeChart in the Java implementation, and will need chart/graphs support in HTML5

@pixelzoom, not a rush, but at some point could you summarize some items that griddle would need to support the needs of Fourier? It might be good to get a feel for if we can chip away at some of those features.

**Comment by pixelzoom** [comment]
Summary of graphs in the Java version of "Fourier: Making Waves".  Recommended to run the sim to get first-hand experience. I don't know what will be added/removed/changed for the HTML5 version, so this may not be complete/correct.

* Bar graph to set amplitudes of Fourier components. Bars are interactive, with adjustable bar height - interacting with the bars changes the component amplitudes. Designers are liable to be specific about how these bars are adjusted, unknown whether this will change in HTML5 version.  I'm not convinced this is general enough for griddle. Example from _Discrete_ screen:

<img width="741" alt="screenshot_400" src="https://user-images.githubusercontent.com/3046552/84943563-c3d59880-b0a1-11ea-80d8-9277dc267b9f.png">

* Plot harmonics and sum as sin waves, fixed or scrolling. When "Function of space & time" is selected, these waves move, and needs to be performant with ~12 plots. Different colors for each plot. Zoom for x and y axes. Example from _Discrete_ screen:

<img width="742" alt="screenshot_404" src="https://user-images.githubusercontent.com/3046552/84944484-11063a00-b0a3-11ea-91d2-f959acbf1d44.png">

* Plot Fourier components as bars, sum as a curve. Show/hide sum. Example from _Discrete to Continuous_ screen:

<img width="740" alt="screenshot_402" src="https://user-images.githubusercontent.com/3046552/84944058-773e8d00-b0a2-11ea-93f8-c8c7f5d0029d.png">

* All graphs use MVC pattern and have a model-view transform, so that "tools" can easily be overlaid on the graphs. E.g. "Wavelength tool" (λ2) and "Period tool" (T<sub>2</sub>) in _Discrete_ screen:

<img width="741" alt="screenshot_406" src="https://user-images.githubusercontent.com/3046552/84945148-20d24e00-b0a4-11ea-8be8-d74d9bf1c178.png">

**Comment by ariel-phet** [comment]
Sounds like there won't be much movement on this topic for Fourier, so closing.

**Comment by pixelzoom** [comment]
Reopening.  In https://github.com/phetsims/fourier-making-waves/issues/3#issuecomment-721293038, I said:

> * griddle’s biggest problem is that it lacks an “overall design”. ...
> ...
> ...  If PhET is willing to incur the additional cost, and an “overall design” for griddle is put in place, consider migrating Fourier charts to griddle at the completion of Fourier implementation.

This issue is about coming up with "big picture", which is the overall design.  Does closing this issue mean that there will be n movement on design, and therefore no migration from Fourier to griddle?

**Comment by pixelzoom** [comment]
I'm also curious about why this issue is being closed because "there won't be much movement on this topic for Fourier".   I've evaluated griddle for 2 sims that make heavy use of charts (Natural Selection, Fourier).  griddle was found lacking in both cases, and we made the (imo unfortunate) decision to proceed with expensive sim-specific charts.   There is a problem here, and this issue is supposed to generally address that problem.  So can you elaborate on why this issue should be closed?  How does PhET plan to address the need for a chart library?

**Comment by pixelzoom** [comment]
Reopening.  This issue was created to address the big picture for griddle, which includes overall design.

In https://github.com/phetsims/fourier-making-waves/issues/3#issuecomment-721293038, I said:

> * griddle’s biggest problem is that it lacks an “overall design”. ...

I'm curious about why this issue is being closed because "there won't be much movement on this topic for Fourier". I've evaluated griddle for 2 sims that make heavy use of charts (Natural Selection, Fourier). griddle was found lacking in both cases, and we made the (imo unfortunate) decision to proceed with expensive sim-specific charts. 

There is a need here that is not being met, and this issue is supposed to generally address that. Can you elaborate on why this issue should be closed, vs deferred? If it is closed, why do we expect the next evaluation of griddle to be different? How does PhET plan to address the need for a chart library? What happens when PhET needs PhET-iO and a11y for charts?

Re the plan to implement Fourier-specific charts, I also said:

> ...  If PhET is willing to incur the additional cost, and an “overall design” for griddle is put in place, consider migrating Fourier charts to griddle at the completion of Fourier implementation.

Closing this issue means that there will be no "overall plan", and therefore no migration from Fourier to griddle.  Is that OK? If so, then I won't put effort into making Fourier charts general, since that will be more expensive.

@ariel-phet can you please clarify?

**Comment by ariel-phet** [comment]
Sounds good @pixelzoom - I will mark deferred instead and this issue can hang out until the time comes when this work is perceived as a priority.

**Comment by samreid** [comment]
Brainstorming some other options regarding Fourier:

* Use some parts of griddle (like the chart backgrounds), even if we do not use everything in griddle
* Any charting parts that are added for Fourier should be added in griddle, with the intent that they can be reused in the future

**Comment by pixelzoom** [comment]
@samreid said:
> * Any charting parts that are added for Fourier should be added in griddle, with the intent that they can be reused in the future

I can't implement something that "should be added in griddle" when I have no idea what the needs or requirements are, and no resources will be allocated to identify those requirements.  If this is a goal, then PhET needs to identify an overall plan for griddle, which at the very least attempts to describe the charting needs for future sims. And my understanding (from comments above) is that resources will not be allocated to griddle during the Fourier timeframe.

**Comment by samreid** [comment]
> we made the (imo unfortunate) decision to proceed with expensive sim-specific charts.

I agree. In my opinion it is shortsighted and inefficient to keep adding sim-specific charts.

> Sounds like there won't be much movement on this topic for Fourier, [...]

@ariel-phet can you please clarify? Is there room for this decision to be re-evaluated? 

UPDATE: I'm removing the deferred label in case @ariel-phet isn't looking at deferred issues, this just means the issue is undeferred for discussion, not undeferred for work.

UPDATE: https://github.com/phetsims/fourier-making-waves/issues/3#issuecomment-723342368 was written after this comment, I'll add notes there and restore this issue to its former status.

**Comment by samreid** [comment]
Bamboo corrects many of the deficiencies identified in this issue, and will be used for development in Fourier, and Eating & Exercise, and is already in use in Circuit Construction Kit: AC.
